feat: Add automatic login flow support

kamilogorek opened the pull request:
Adds automated login flow, which uses the browser session to authenticate and generate a new token, which is used in the cli `login` command.

```
Usage:
  supabase login [flags]

Flags:
  -h, --help           help for login
      --name string    Name that will be used to store token in your settings, defaults to built-in token name generator
      --no-browser     Do not open browser automatically
      --token string   Use provided token instead of automatic login flow
```

RFC: https://www.notion.so/supabase/RFC-Automatic-browser-based-CLI-login-flow-fc4f91fc7e7f4f168d2da6192841dcec?d=d8c9bcbe826c4665a89f9240d0854a36

API changes: https://github.com/supabase/infrastructure/pull/15529
Studio changes: https://github.com/supabase/supabase/pull/18641

sweatybridge commented on internal/login/login.go:
```diff
@@ -1,18 +1,185 @@
+		fmt.Printf("Here is your login link: %s\n\n", utils.Bold(createLoginSessionUrl))
+	}
+
+	spinner := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
```
We have been using bubbletea/spinner for existing commands like [supabase stop](https://github.com/supabase/cli/blob/main/internal/stop/stop.go#L26). Could you try if that works?

kamilogorek replied:
Updated. Although it's now quite obvious why I went with other library for such a simple UI thing :p

https://github.com/supabase/cli/pull/1630/commits/127b9e2723f7294a69e7a4db1892203639155ff3#diff-63f7df4adcd885e26af5017d4653902be40956c5d99878eb9795b1c406b16d77

kamilogorek replied:
Tbh we can also remove it entirely, or redo `RunProgram` to return the value instead of going through channels

sweatybridge commented on internal/login/login.go:
```diff
@@ -1,18 +1,185 @@
+	var accessTokenResponse AccessTokenResponse
+
+	// We fully control the url here, so it's safe to perform the request and ignore the G107 gosec rule.
+	resp, err := http.Get(url) //nolint:gosec
```
Once backend moves to [v1 route](https://github.com/supabase/infrastructure/pull/15529/files#r1381779789), it would be nice to replace this with the http client generated from openapi spec.

sweatybridge commented on internal/login/login.go:
```diff
@@ -1,18 +1,185 @@
 )
 
-func Run(stdin *os.File, fsys afero.Fs) error {
-	accessToken := PromptAccessToken(stdin)
```
Are we going to stop prompting for access token completely? This could be a breaking change for users who pipe in tokens from CI environments, for eg. `supabase login <<< "token"`.

This is probably a rare case but I'd prefer keeping backwards compatibility if we can.

kamilogorek replied:
Good catch, will update.

sweatybridge commented on cmd/login.go:
```diff
@@ -3,27 +3,77 @@ package cmd
 
 func init() {
+	loginFlags := loginCmd.Flags()
+	loginFlags.String("token", "", "Use provided token instead of automatic login flow")
```
Passing in the token via cli flag seems insecure because it could be saved in shell history. Shall we only support `SUPABASE_ACCESS_TOKEN` env var?

kamilogorek replied:
I used `--auth-token` flags in the past extensively and not a single person reported this as an issue tbh

sweatybridge commented on internal/login/login.go:
```diff
@@ -1,18 +1,225 @@
+		fmt.Printf("Here is your login link: %s\n\n", utils.Bold(createLoginSessionUrl))
+	}
+
+	pollingProcess := tea.NewProgram(pollingState{
+		message: "Your token is now being generated and securely encrypted. Waiting for it to arrive...",
+		spinner: utils.CreateSpinner(),
+	})
+
+	tokenCh := make(chan AccessTokenResponse, 1)
+	errCh := make(chan error, 1)
+
+	go func() {
+		sessionPollingUrl := utils.GetSupabaseAPIHost() + "/platform/cli/login/" + sessionId
+		accessTokenResponse, err := pollForAccessToken(sessionPollingUrl)
+		if err != nil {
+			errCh <- err
+		} else {
+			tokenCh <- accessTokenResponse
+		}
+		pollingProcess.Quit()
+	}()
+
+	if _, err := pollingProcess.Run(); err != nil {
+		return fmt.Errorf("cannot render polling status: %w", err)
+	}
+
+	select {
+	case err := <-errCh:
+		return err
+	case accessTokenResponse := <-tokenCh:
+		decryptedAccessToken, err := decryptAccessToken(accessTokenResponse, curve, privateKey)
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("Token created successfully: %s\n\n", utils.Bold(tokenName))
+		fmt.Println(loggedInMsg)
+
+		return utils.SaveAccessToken(decryptedAccessToken, params.Fsys)
+	}
```
```suggestion
	return utils.RunProgram(ctx, func(p utils.Program, ctx context.Context) {
		p.Send(utils.StatusMessage("Your token is now being generated and securely encrypted. Waiting for it to arrive..."))

		sessionPollingUrl := utils.GetSupabaseAPIHost() + "/platform/cli/login/" + sessionId
		accessTokenResponse, err := pollForAccessToken(sessionPollingUrl)
		if err != nil {
			return err
		}
		decryptedAccessToken, err := decryptAccessToken(accessTokenResponse, curve, privateKey)
		if err != nil {
			return err
		}

		fmt.Printf("Token created successfully: %s\n\n", utils.Bold(tokenName))
		fmt.Println(loggedInMsg)

		return utils.SaveAccessToken(decryptedAccessToken, params.Fsys)
	})
```

I think something like this would work? Not much changes given our current refactoring.

kamilogorek replied:
Perfection :chefs_kiss:

kamilogorek replied:
Had to wait for program to end first, otherwise prints overlap, as bubbletea controls terminal state inside the program.

sweatybridge commented on internal/login/login.go:
```diff
@@ -1,18 +1,225 @@
+	if params.OpenBrowser {
+		fmt.Printf("Here is your login link in case browser did not open: %s\n\n", utils.Bold(createLoginSessionUrl))
+
+		openCmd := exec.Command("open", createLoginSessionUrl)
```
```suggestion
		openCmd := exec.CommandContext(ctx,
"open", createLoginSessionUrl)
```

Use context to allow cancellation.

kamilogorek commented on internal/login/login.go:
```diff
@@ -1,18 +1,193 @@
+	sessionId := uuid.New().String()
+	tokenName := params.Name
+	curve := ecdh.P256()
+	privateKey, err := curve.GenerateKey(rand.Reader)
```
I talked with Stojan about it, and he said that it’s best to just let the session expire on the backend and force user to retry. It should rarely happen, and it will work just fine if they retry the command.

sweatybridge replied:
Ok, I haven't read the dashboard code. How do we deal with dangling access tokens that were left behind from cancelled or interrupted login sessions?

kamilogorek replied:
They stay assigned to your account and you can delete them manually.
I'll add a note about where they are stored alongside the link to settings page from the "success" login page, good idea.

kamilogorek commented:
Let's merge and wait for API changes deployment, and once it's done, I'll confirm that the CLI is working as expected, and we'll be able to throw it into the beta release newt.

sweatybridge commented:
Our beta release happens on merge. So the order should be first deploy api changes to staging, test with cli, then merge to cli.

kamilogorek commented:
That was exactly my thinking, bad wording, sorry :D

sweatybridge commented on internal/login/login.go:
```diff
@@ -1,18 +1,193 @@
+	return string(decryptedAccessToken), nil
+}
+
+func pollForAccessToken(url string) (AccessTokenResponse, error) {
```
Let's add more unit tests for this. It would help with future refactoring if we change the http client.

You can check the uncovered lines here https://coveralls.io/builds/63814279/source?filename=internal%2Flogin%2Flogin.go#L90

kamilogorek commented:
Added one test that covers the "happy path", as it's quite complicated to test this whole feature in the first place. Had to jump through some hoops 👀 
Also fixed broken `sigint` handling when spinner was in its pending state.
